### PR TITLE
[CSM-224] Store address balances for Explorer

### PR DESCRIPTION
**Pull request by Pastafarianist**
Continuation of https://github.com/input-output-hk/cardano-sl/pull/297 with a changed base branch.

### CODE REVIEW

File: src/Pos/Explorer/Txp/Init.hs [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+
+initializeGenesisBalances :: (WithNodeContext ssc m, MonadDB m) => m ()
+initializeGenesisBalances =
+    unlessM hasInitMarker $ do
+        genesisUtxo <- genesisUtxoM
+        putGenesisBalances genesisUtxo
+        putInitMarker
+
+initMarker :: ByteString
+initMarker = "e/init"
```
Comment (gromakovsky) [inline src/Pos/Explorer/Txp/Init.hs]: We usually put `/` to the end of each key to decrease probability of having one key as a prefix of another one.

File: src/Pos/DB/DB.hs [hunk 2]
```diff
@@ -87,20 +92,9 @@ initNodeDBs = do
...
-    => m (Block ssc)
-getTipBlock = maybe onFailure pure =<< getBlock =<< getTip
-  where
-    onFailure = throwM $ DBMalformed "there is no block corresponding to tip"
-
--- | Get BlockHeader corresponding to tip.
-getTipBlockHeader
-    :: (SscHelpersClass ssc, MonadDB m)
-    => m (BlockHeader ssc)
-getTipBlockHeader = getBlockHeader <$> getTipBlock
```
Comment (gromakovsky) [inline src/Pos/DB/DB.hs]: Why did you move these functions?

File: src/Pos/Explorer/Txp/Toil/Logic.hs [hunk 3]
```diff
@@ -192,10 +193,15 @@ updateAddrBalances (combineBalanceUpdates -> updates) = mapM_ updater updates
...
-            Nothing -> logError "updateAddrBalances: subtracting coins from unknown address"
+            Nothing ->
+                logError $
+                    sformat ("updateAddrBalances: attempted to subtract "%build%" coins from unknown address "%build)
+                    coin addr
             Just currentBalance
                 | currentBalance < coin ->
-                    logError "updateAddrBalances: address does not have enough coins"
+                    logError $
+                        sformat ("updateAddrBalances: attempted to subtract "%build%" coins from address "%build%" which only has "%build%" coins")
```
Comment (gromakovsky) [inline src/Pos/Explorer/Txp/Toil/Logic.hs]: This line seems to be extremely long 😔 

Also, coin is formatted as `5 coin(s)`. If you write `coins` in format, then use `int` formatter (which will format it as `5`).

File: src/Pos/Explorer/Txp/Init.hs (hunk 1 above)
Comment (Pastafarianist) [inline src/Pos/Explorer/Txp/Init.hs]: I'm looking at https://github.com/input-output-hk/cardano-sl/blob/2ea6c88d06167c2581a22e3e2e4e3e59d2b9c03c/src/Pos/Txp/DB/Utxo.hs#L212 and it's not there. Should it be added there as well?

File: src/Pos/DB/DB.hs (hunk 2 above)
Comment (gromakovsky) [inline src/Pos/DB/DB.hs]: Well, `Pos.DB` shouldn't import `Pos.Explorer`, actually nothing should import `Pos.Explorer` and it should be moved to `cardano-sl-explorer`. But it's hard to do now, so let's forget about it for now.

File: src/Pos/Explorer/Txp/Init.hs (hunk 1 above)
Comment (gromakovsky) [inline src/Pos/Explorer/Txp/Init.hs]: I also noticed that. I think it should be added at least for consistency.

File: src/Pos/Explorer/DB.hs [hunk 4]
```diff
@@ -29,6 +38,38 @@ getAddrHistory :: MonadDB m => Address -> m AddrHistory
...
+
+areBalancesInitialized :: MonadDB m => m Bool
+areBalancesInitialized = isJust <$> (getUtxoDB >>= rocksGetBytes balancesInitFlag)
+
+putInitFlag :: MonadDB m => m ()
+putInitFlag = gsPutBi balancesInitFlag True
+
+putGenesisBalances :: MonadDB m => Utxo -> m ()
+putGenesisBalances genesisUtxo = do
+    let txOuts = map (view _TxOut . toaOut . snd) . M.toList $ genesisUtxo
```
Comment (gromakovsky) [inline src/Pos/Explorer/DB.hs]: FYI: you could use `toList` instead of `M.toList` and not use `snd`.

Comment (int-index) [inline src/Pos/Explorer/DB.hs]: There's also `M.elems`.

**Comment by akegalj** [comment]
@gromakovsky 
> Test

this is from https://serokell.slack.com/archives/G54P1SQCS/p1494506136102102 

Well, CSM-224 seems to work. Here's what I've done:

```bash
# Assuming that the cardano-sl repository is located in /home/ser/repos/cardano-sl/

PUBLIC_KEY_ADDRESS=1fbPUqmdG1PdYpKxhw8qYc5hC342W3STosbcZNMqsadodxL
REDEEM_ADDRESS=3mbUdEWUU8HFA937K1C352a9qDiu4gsPXkYkuGncVFLmF6

cd /home/ser/repos/cardano-sl/
stack exec cardano-keygen -- --dump-dev-genesis-keys keys/{}.key
echo "\"/home/ser/repos/cardano-sl/keys/1.key\"" | http POST localhost:8090/api/wallets/keys/
http localhost:8100/api/addresses/summary/$(PUBLIC_KEY_ADDRESS)
http POST localhost:8090/api/txs/payments//$(PUBLIC_KEY_ADDRESS)/$(REDEEM_ADDRESS)/124
http localhost:8100/api/addresses/summary/$(REDEEM_ADDRESS)
```

Output:

```bash
HTTP/1.1 200 OK
Content-Type: application/json;charset=utf-8
Date: Thu, 11 May 2017 12:11:55 GMT
Server: Warp/3.2.11.1
Transfer-Encoding: chunked

{
    "Right": {
        "caAddress": "3mbUdEWUU8HFA937K1C352a9qDiu4gsPXkYkuGncVFLmF6",
        "caBalance": {
            "getCoin": "124"
        },
        "caTxList": [
            {
                "ctbId": "be61b4dbc980690b46157e4471f53fc16b81d4aeb2e6c5ed1947c082a3f0b238",
                "ctbInputSum": {
                    "getCoin": "33333"
                },
                "ctbInputs": [
                    [
                        "1fbPUqmdG1PdYpKxhw8qYc5hC342W3STosbcZNMqsadodxL",
                        {
                            "getCoin": "33333"
                        }
                    ]
                ],
                "ctbOutputSum": {
                    "getCoin": "33333"
                },
                "ctbOutputs": [
                    [
                        "1fbPUqmdG1PdYpKxhw8qYc5hC342W3STosbcZNMqsadodxL",
                        {
                            "getCoin": "33209"
                        }
                    ],
                    [
                        "3mbUdEWUU8HFA937K1C352a9qDiu4gsPXkYkuGncVFLmF6",
                        {
                            "getCoin": "124"
                        }
                    ]
                ],
                "ctbTimeIssued": 1494504679.052242
            }
        ],
        "caTxNum": 1,
        "caType": "CRedeemAddress"
    }
}
```

So the address summary endpoint is able to return info on `RedeemAddress`es.

File: src/Pos/Explorer/Txp/Toil/Logic.hs [hunk 5]
```diff
@@ -95,8 +96,9 @@ eProcessTx tx@(id, aux) extra = do
     undo <- Txp.processTx tx
     putTxExtra id extra
     putNewHistory id $ getTxRelatedAddrs aux undo
-    -- let balanceUpdate = getBalanceUpdate aux undo
-    -- updateAddrBalances balanceUpdate
+    let balanceUpdate = getBalanceUpdate aux undo
+    fmap fst $ usingLoggerName "eProcessTx" $ runNamedPureLog $
```
Comment (gromakovsky) [inline src/Pos/Explorer/Txp/Toil/Logic.hs]: Here log events are discarded. Please add a TODO and also it makes sense to create a minor issue about it (in this case put issue identifier into TODO).

Comment (Pastafarianist) [inline src/Pos/Explorer/Txp/Toil/Logic.hs]: Done.

**Comment by Pastafarianist** [comment]
Local txp is still broken, but this isn't a fault of this PR. Logged as CSM-246.
